[Lee Shao Wee] IP

shaowi opened the pull request:
# Shao

> “In order to be irreplaceable, one must always be different” – Coco Chanel ([Source](https://apiumhub.com/tech-blog-barcelona/programming-jokes-quotes/))

Shao is an easy to use chat application that allows you to:

- Record down the things you need **to do**.
- Set a _deadline_ for something you ~~plan to do~~ already planned to do.
- Make an **_event_** for an occasion with it's start and end date.

This app is/can:

1. **Easy** to use
2. **Free** to use
3. **Become** your new best friend :wink:

To run the program, make sure that you have:

- [ ] A Windows/Mac operating system
- [ ] An IDE installed
- [ ] JDK 19

Here is a snippet of the `main` method:

```Java
public class Shao {
    public static void main(String[] args) {
        new Shao().run(args);
    }
}
```

Commit:
Add Gradle support

shaowi committed:
initial commit

shaowi committed:
modify git ignore

shaowi committed:
Level-1

shaowi committed:
Level-2

shaowi committed:
Level-3

shaowi committed:
Level-4

shaowi committed:
A-TextUiTesting

shaowi committed:
Level-5

shaowi committed:
Level-6

shaowi committed:
A-Enums

shaowi committed:
Update runtest.sh

shaowi committed:
Remove docs folder

shaowi committed:
Add docs folder to gitignore

shaowi committed:
Updated README

shaowi committed:
Add text file to save records

shaowi committed:
Read saved data and save new added task

shaowi committed:
Save mark and unmarked items status

shaowi committed:
Delete saved tasks

shaowi committed:
Fulfilling minimum requirements

shaowi committed:
Stretch goal:

Add a command "datetime (datetime)" to show the list of tasks that falls
on specified datetime

shaowi committed:
Merge branch 'branch-Level-8'

shaowi committed:
Fix merge conflicts
- Update class constructors data types
- Solve issues pertaining to file inputs

shaowi committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

shaowi committed:
A-Gradle

shaowi committed:
Place java files into seedu folder

shaowi committed:
Updated README

shaowi committed:
Fix 'NoSuchElement' bug issues

shaowi committed:
Add vscode to gitignore

shaowi committed:
Setup unit test
- Modify build.gradle to point to the class source file
- Add TodoTest and EventTest

shaowi committed:
Include fixes from Level-8

shaowi committed:
A-JUnit
- Continue writing tests for Event
- Add test to Shao
- Modify method inputs

shaowi committed:
Make the code more OOP
- Refactored code by separating them to folders ui, task, storage and
parser

shaowi committed:
Moved and renamed fetchData method from Storage
- Changed to parseInput in Parser

shaowi committed:
Extracted to command classes
- Refactor code to parse command from input

shaowi committed:
Updated runtest to compile all sub folders java file

shaowi committed:
Deleted all test files

shaowi committed:
Fix mark, add, delete task operation

shaowi committed:
Refactor to fix compilation issues

shaowi committed:
Update gitignore

shaowi committed:
Modify expected file inputs

shaowi committed:
Add test files

shaowi committed:
Update runtest.sh to be consistent with run test.bat

shaowi committed:
Modify root class name

shaowi committed:
Added JUnit Tests
- Parser and task modules

shaowi committed:
Remove unnecessary comments

shaowi committed:
Add JavaDoc comments

shaowi committed:
Follow the Coding Standard

shaowi committed:
Give users a way to find a task by searching for a keyword

shaowi committed:
Modify test input

shaowi committed:
Merge branch 'branch-A-CodingStandard'

shaowi committed:
Refactor method

shaowi committed:
Merge branch 'branch-A-CodingStandard'

shaowi committed:
Merge branch 'branch-Level-9'

shaowi committed:
Add JavaDoc for a method

shaowi commented:
# Shao

> “In order to be irreplaceable, one must always be different” – Coco Chanel ([Source](https://apiumhub.com/tech-blog-barcelona/programming-jokes-quotes/))

Shao is an easy to use chat application that allows you to:

- Record down the things you need **to do**.
- Set a _deadline_ for something you ~~plan to do~~ already planned to do.
- Make an **_event_** for an occasion with it's start and end date.

This app is/can:

1. **Easy** to use
2. **Free** to use
3. **Become** your new best friend :wink:

To run the program, make sure that you have:

- [ ] A Windows/Mac operating system
- [ ] An IDE installed
- [ ] JDK 19

Here is a snippet of the `main` method:

```Java
public class Shao {
    public static void main(String[] args) {
        new Shao().run(args);
    }
}
```

shaowi committed:
Updated README

shaowi committed:
Added avatar image files

shaowi committed:
Initialise JavaFx

shaowi committed:
Modified command to adapt GUI functions
- Welcome message
- Unknown inputs

shaowi committed:
Add *.txt files

shaowi committed:
Modified program functions from CLI to GUI
- Styles for output components
- Update commands to receive input and send output in GUI

shaowi committed:
Modify error message on datetime parse fail
- DatetimeException when parsing datetime string should be handled with an appropriate error message

shaowi committed:
Update deadline and event date conditional checks
- /to should be later or equal to /from
- /from should be later than current datetime

shaowi committed:
Updated JavaDoc

shittake commented on src/main/java/Shao.java:
```diff
@@ -0,0 +1,153 @@
+import parser.Parser;
+import storage.Storage;
+import tasklist.TaskList;
+import ui.Ui;
```
Perhaps you could leave a line between different packages imported for better readability?

shittake commented on src/main/java/commands/AddCommand.java:
```diff
@@ -0,0 +1,138 @@
+						: TaskType.EVENT;
+	}
+
+	/**
+	 * Set task from task type.
+	 * 
+	 * @param inputArr
+	 * @param ui
+	 * @param parser
+	 * @param storage
+	 * @param dialogContainer
+	 * @return boolean
+	 */
```
I think it would be clearer if you could briefly explain what each parameter is for each JavaDoc documentation

shittake commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,244 @@
+		}
+
+		try {
+			if (dateTimeStr.contains("/")) { // Input Format: dd/MM/YYYY
+				dateArr = Stream.of(dateTimeArr[0].split("/"))
+						.map(Integer::valueOf)
+						.toArray(Integer[]::new);
+				return LocalDateTime.of(LocalDate.of(dateArr[2], dateArr[1], dateArr[0]), time);
```
I like how you chain stream methods in this manner which makes it look neat and aligned

shittake commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,244 @@
+	 * @return LocalDateTime
+	 */
+	public LocalDateTime getBy(String[] inputArr, Ui ui, Storage storage, VBox dialogContainer) {
+		int l = inputArr.length;
+		for (int i = 0; i < l; i++) {
+			if (i < l - 1 && inputArr[i].equals("/by")) {
+				return parseDateTimeStr(sliceArrAndConcate(inputArr, i + 1, l), ui, storage, dialogContainer);
+			}
+		}
```
Perhaps use a more meaningful variable name instead of l? Since it is in a larger scope than variable i, I believe it would be better to use something else like `arrLength`

shittake commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,244 @@
+			default:
+				break;
+		}
+		if (inputArr[1].equals("1"))
+			newTask.markAsDone();
```
To fit the coding standards, you could use opening and closing braces for this `if` condition

shittake commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,244 @@
+	public String trimDate(String input) {
+		String[] inputArr = input.split(" ");
+		StringBuilder sb = new StringBuilder();
+		for (String s : inputArr) {
+			if (s.contains("/")) {
+				return sb.toString().trim();
+			}
+			sb.append(s + " ");
```
I believe a more meaningful variable name for each String could be used instead of `s`, such as `untrimmedDateStr`

shittake reviewed (commented):
Looks good to merge!

Great job 👍

shaowi committed:
Generate a new JAR file
- Rename files to fix error in packing javaFX dependencies

shaowi committed:
Organise module imports: add row space between different imported
module packages

Modules imported are lump together as one section.

Hard to differentiate modules imported from different packages.

Leave a line between different imported module packages.

A line break separates them and improves readability.

shaowi committed:
Add description to JavaDocs @param

All the parameters does not have a description.

Other developers may not know the what the parameters is for/about.

Included a short description of the parameters passed into a method.

Keeping a clear and concise description will not flood the docs.

shaowi committed:
Add assert on targeted methods

shaowi committed:
Improve code quality

- Magic numbers are extracted to variables
- Extracted common behaviour methods into class methods.

shaowi committed:
Merge pull request #1 from shaowi/branch-A-Assertions

Add assert on targeted methods

shaowi committed:
Merge pull request #2 from shaowi/branch-A-CodeQuality

Improve code quality

shaowi committed:
Add extension to detect and disallow duplicate tasks

nicljr commented on src/main/java/commands/AddCommand.java:
```diff
@@ -0,0 +1,154 @@
+		switch (taskType) {
+			case TODO:
+				if (tasklist.hasDuplicateTodo(description)) {
+					ui.sendResponse(dialogContainer, storage, ui.createLabel(duplicatedErrorMessage));
```
Perhaps this line could be split into two separate lines

nicljr commented on src/main/java/commands/AddCommand.java:
```diff
@@ -0,0 +1,154 @@
+				}
+				if (tasklist.hasDuplicateEvent(description, from, to)) {
+					errorMessage = duplicatedErrorMessage;
+					ui.sendResponse(dialogContainer, storage, ui.createLabel(duplicatedErrorMessage));
```
Same here I think you can split this into two lines so as to keep the line length short

nicljr commented on src/main/java/commands/AddCommand.java:
```diff
@@ -0,0 +1,154 @@
+
+public class AddCommand extends Command {
+
+	enum TaskType {
```
I like the use of enums here for the different types of Task!

nicljr commented on src/main/java/commands/AddCommand.java:
```diff
@@ -0,0 +1,154 @@
+	 */
+	private boolean setTask(String[] inputArr, Ui ui, Parser parser, Storage storage, VBox dialogContainer,
+			TaskList tasklist) {
+		String description = parser.sliceArrAndConcate(inputArr, 1, inputArr.length);
```
I like the naming of the parser function as it describes very clearly its purpose and the intended output!

nicljr reviewed (commented):
Overall, your code follows the coding standard and I like your code structure! Keep up the good work!

shaowi committed:
Moved images to resources folder and added an app icon

shaowi committed:
Delete unused images

shaowi committed:
Updated README with image of app

shaowi committed:
Set up product website
- Added a representative screenshot of the product
- Add a brief User Guide(UG) to Shao

shaowi committed:
Remove docs folder from gitignore

shaowi committed:
Reverted 'moved images to resources folder'

shaowi committed:
Updated test files to pass tests

shaowi committed:
Fix CTFontLogSystemFontNameRequest error

shaowi committed:
Modify image path for user and bot

shaowi committed:
Remove assets/image
